lib/streamlit/server/Server.py

```diff
@@ -260,57 +261,66 @@ def browser_is_connected(self):
 
     @tornado.gen.coroutine
     def _loop_coroutine(self, on_started=None):
-        if self._state == State.INITIAL:
-            self._set_state(State.WAITING_FOR_FIRST_BROWSER)
-        elif self._state == State.ONE_OR_MORE_BROWSERS_CONNECTED:
-            pass
-        else:
-            raise RuntimeError("Bad server state at start: %s" % self._state)
-
-        if on_started is not None:
-            on_started(self)
-
-        while not self._must_stop.is_set():
-            if self._state == State.WAITING_FOR_FIRST_BROWSER:
-                pass
-
+        try:
```

Comment:
For readability, can you pull this whole "try" block into its own function? This way `_loop_coroutine()` would become:

```
@tornado.gen.coroutine
def _loop_coroutine(self, on_started=None):
  try:
    self._real_loop(on_started)
  except:
    # ...
  finally:
    # ...
```

...which is much more readable.

Reply:
I'm hitting a strange situation here, `traceback` module and exceptions in Tornado's `coroutines` don't get along very well, after creating the wrapper function the traceback prints out the coroutine exception, although if I print the exception itself the error I wanted it there. 

This is what it gets printed out, the first line is the exception from the `print` line, the rest is what is being printed out by `traceback` module.

```
EXCEPTION! 'Server' object has no attribute 'asdasd'
  File "/Users/dnuske/venvs/lib-OhQ5BDDw/bin/streamlit", line 11, in <module>
    load_entry_point('streamlit', 'console_scripts', 'streamlit')()
  File "/Users/dnuske/venvs/lib-OhQ5BDDw/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/dnuske/venvs/lib-OhQ5BDDw/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/dnuske/venvs/lib-OhQ5BDDw/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/dnuske/venvs/lib-OhQ5BDDw/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/dnuske/venvs/lib-OhQ5BDDw/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/dnuske/code/streamlit/streamlit/lib/streamlit/cli.py", line 154, in main_run
    _main_run(file_or_url, args)
  File "/Users/dnuske/code/streamlit/streamlit/lib/streamlit/cli.py", line 179, in _main_run
    bootstrap.run(file, command_line, args)
  File "/Users/dnuske/code/streamlit/streamlit/lib/streamlit/bootstrap.py", line 186, in run
    ioloop.start()
  File "/Users/dnuske/venvs/lib-OhQ5BDDw/lib/python3.7/site-packages/tornado/platform/asyncio.py", line 132, in start
    self.asyncio_loop.run_forever()
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py", line 534, in run_forever
    self._run_once()
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/base_events.py", line 1771, in _run_once
    handle._run()
  File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/Users/dnuske/venvs/lib-OhQ5BDDw/lib/python3.7/site-packages/tornado/ioloop.py", line 758, in _run_callback
    ret = callback()
  File "/Users/dnuske/venvs/lib-OhQ5BDDw/lib/python3.7/site-packages/tornado/stack_context.py", line 300, in null_wrapper
    return fn(*args, **kwargs)
  File "/Users/dnuske/venvs/lib-OhQ5BDDw/lib/python3.7/site-packages/tornado/gen.py", line 346, in wrapper
    runner = Runner(result, future, yielded)
  File "/Users/dnuske/venvs/lib-OhQ5BDDw/lib/python3.7/site-packages/tornado/gen.py", line 1080, in __init__
    self.run()
  File "/Users/dnuske/venvs/lib-OhQ5BDDw/lib/python3.7/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/Users/dnuske/code/streamlit/streamlit/lib/streamlit/server/Server.py", line 273, in _loop_coroutine_wrapper
    traceback.print_stack(file=sys.stdout)
```

One option is to print out the traceback straight out of the `Exception`, the other one is going back to not having the wrapper function, and I read about the co-routine returning the exception as a `Future` object, have to take a look at that, it is a possibility.